### PR TITLE
Basic infra for annotate shapes and dtypes on arguments.

### PR DESCRIPTION
**Pull request by silvasean**
These allow users to annotate a known "type bound" on the argument,
which can seed shape/dtype inference. We don't rewrite the function
types as part of the import process (it will happen in a
yet-to-be-written pass) because:

1. We would need to interprocedurally rewrite all calls to keep the IR
   consistent. Currently, we have a place after GlobalizeObjectGraph but
   before we convert to tensors where this is convenient to do. Ideally,
   we would do this on the object graph representation.

1. We don't necessarily know that adjusting the function type is a legal
   calling convention change. The pass will have blessed knowledge (by
   the pass pipeline author) that adjusting the argument type based on
   the type bound is safe (which it frequently is).

2. Note that in principle, a type bound could be a fairly general thing
   (such as maximum sizes of dimensions, unions of multiple concrete
   types, etc.). The pass will in principle have logic to interpret the
   type bounds and to determine a suitable "best" (and legal) argument
   type.

### CODE REVIEW

File: frontends/pytorch/csrc/builder/class_annotator.cpp [hunk 1]
```diff
@@ -152,10 +141,115 @@ ClassAnnotator::getOrCreateClassAnnotation(c10::ClassType *classType) {
...
+          &it->second->getMethodAnnotations()[i];
+    }
   }
   return *it->second;
 }
 
+static c10::ScalarType convertToC10ScalarType(py::object obj) {
+  if (THPDtype_Check(obj.ptr())) {
+    // Need C-style cast, since no C++-level inheritance is involved.
+    THPDtype *dtype = (THPDtype *)(obj.ptr());
```
Comment (stellaraccident) [inline frontends/pytorch/csrc/builder/class_annotator.cpp:155]: Could be `reinterpret_cast`, right?

File: frontends/pytorch/csrc/builder/class_annotator.h [hunk 2]
```diff
@@ -88,6 +111,9 @@ using ClassAnnotationMap =
     std::unordered_map<c10::ClassType *, std::unique_ptr<ClassAnnotation>>;
 
 // A collection of class annotations + methods to create the annotations.
+//
+// This object is bound into Python, but the UI is quite poor. We expect
+// some amount of Python metaprogramming syntax sugar to make it usable.
```
Comment (stellaraccident) [inline frontends/pytorch/csrc/builder/class_annotator.h:116]: Metaprogramming. My favorite kind of programming ;)

File: frontends/pytorch/test/ivalue_import/annotations/export-recursive.py [hunk 3]
```diff
@@ -41,8 +41,8 @@ def forward(self):
 # CHECK:           torch.method private "not_exported_method", @{{.*}}
 # CHECK:         }              
 annotator.exportNone(class_type)
-annotator.exportPath(['s', 'exported'], class_type)
-annotator.exportPath(['s', 'forward'], class_type)
+annotator.exportPath(class_type, ['s', 'exported'])
```
Comment (stellaraccident) [inline frontends/pytorch/test/ivalue_import/annotations/export-recursive.py:44]: This is making my eyes bleed. I know you want to sugar it later, but wdyt about adding an overload to `exportPath` that takes a string arg and splits on dots?

(also nit: these should probably be snake-case)

**Review by stellaraccident — APPROVED** [review]

File: frontends/pytorch/test/ivalue_import/annotations/export-recursive.py (hunk 3 above)
Comment (silvasean) [reply frontends/pytorch/test/ivalue_import/annotations/export-recursive.py:44]: Is there an easy way to add that in Python? I don't want to pollute the C++/Python interface with sugar-only methods.

I could also switch the methods to always take a dotted string. I guess that would be robust enough too. WDTY?

Comment (silvasean) [reply frontends/pytorch/test/ivalue_import/annotations/export-recursive.py:44]: Will fix in a future patch.

**Pull request merged**